### ISSUE TITLE
Consistency with Torch.utils.data.DataLoader

### ISSUE THREAD

**Issue by 4fee8fea** [NVIDIA/DALI#3750]
After migrating the training pipeline from PyTorch's `DataLoader` to `NVIDIA/DALI`, I wanted to compare the consistency of the loading process between the two. In other words, making sure that the images loaded by `NVIDIA/DALI` were the same as the images loaded by the DataLoader.

The following are my scripts:
```python
import torch
from torch.utils.data import DataLoader
from torch.utils.data.sampler import SequentialSampler

from nvidia.dali import pipeline_def, fn
from nvidia.dali.plugin.pytorch import DALIClassificationIterator
from nvidia.dali.plugin.base_iterator import LastBatchPolicy

class TrainDataset(MyDataset):
    def __init__(self, args):
        super().__init__(args)
    def __getitem__(self, idx):
        patch = self.patches[idx]
        label = self.labels[idx]
        return patch, label

train_loader = DataLoader(train_dataset, batch_size=args.batch_size, shuffle=False, sampler=SequentialSampler(train_dataset), drop_last=True, num_workers=args.num_workers, pin_memory=True, persistent_workers=True)
view = None
for X, label in train_loader:
    view = torch.cat((view, X), dim=0) if view is not None else X

@pipeline_def
def TrainPipeline(args):
    data_root = args.dali_file_root
    data_files = sorted(os.listdir(data_root))
    X = fn.readers.numpy(device='gpu', file_root=data_root, files=data_files, shuffle_after_epoch=False, seed=args.seed)

    return X, X

train_pipeline = TrainPipeline(batch_size=args.batch_size, num_threads=args.num_threads, device_id=0, seed=args.seed, args=args)
train_pipeline.build()
dali_loader = DALIClassificationIterator(pipelines=train_pipeline, auto_reset=True, size=args.size, last_batch_policy=LastBatchPolicy.DROP, last_batch_padded=True)
view_prime = None
for data in dali_loader:
    view_prime = torch.cat((view_prime, data[0]['data']), dim=0) if view_prime is not None else data[0]['data']


view, view_prime = view.cpu(), view_prime.cpu()
torch.allclose(view, view_prime) #output: False
```

I checked that the order of the `*.npy` files read by `fn.readers.numpy` is consistent with the order of the data in the DataLoader (both in `C-format`).

I expect the output of `torch.allclose` to be True, but it is not. What could be the reason for this? Thanks for your time!

**Comment by JanuszL** [comment]
Hi @4fee8fea,

It is not expected. Can you provide a self-contained reproduction script we can run on our side so we can reproduce the problem?

**Comment by 4fee8fea** [comment]
In my first script, `shuffle_after_epoch=False` was set, but the output for `torch.allclose(view, view_prime)` is `False`. I can't recover the situation now. The reason is likely due to my carelessness at that time. 

However, the current test did find that `shuffle_after_epoch` has an impact on the behavior. I think the current results can be trusted.

**Comment by JanuszL** [comment]
Hi @4fee8fea,

This is expected as:
```
shuffle_after_epoch (bool, optional, default = False) – If set to True, the reader shuffles the entire dataset after each epoch.
```
just enables shuffling and the order of samples is randomized.

**Comment by 4fee8fea** [comment]
Now I know it will be shuffled whenever called, LOL

**Comment by songyuc** [comment]
Hi, @JanuszL!
Is `shuffle_after_epoch` in fn.readers a most recommended way to implement `torch.utils.data.DataLoader(..., shuffle=True, ...)` equivalently?
Your answer and guide will be appreciated!

**Comment by JanuszL** [comment]
Hi @songyuc,

That is correct - `shuffle_after_epoch` provides the best level of entropy. Still, in many cases the regular shuffling is sufficient.

**Comment by songyuc** [comment]
> Hi @songyuc,
> 
> That is correct - `shuffle_after_epoch` provides the best level of entropy. Still, in many cases the regular shuffling is sufficient.

Thanks sincerely!
